### PR TITLE
Add nanochat2 model: complete nanochat architecture

### PR DESCRIPTION
**Pull request by olii-dev**
## Requirements

- I understand it is strictly prohibited to use AI to write PR descriptions.
- AI usage disclosure: This description and the implementation were drafted with AI assistance; the author (human, @olii-dev) reviewed both, ran the verification below, and stands behind the content and the code.

## Model information

- Name: `nanochat2` (Lattice Quark 1.5B, a chat model in the nanochat lineage)
- Homepage: https://huggingface.co/lattice-research
- Reference implementation: this architecture is NOT reimplemented in `transformers`. The reference is custom PyTorch training code: https://github.com/olii-dev/quark-5090-q2 (`nanochat/gpt.py`), itself descended from https://github.com/karpathy/nanochat. The existing `nanochat` type in this repo is an incomplete port; `nanochat2` is the full forward pass, matching that torch reference exactly.

## Supported checkpoints

- https://huggingface.co/lattice-research/lattice-quark-1.5b-mlx (4-bit MLX, `model_type: nanochat2`)
- Source (torch) weights: https://huggingface.co/lattice-research/lattice-quark-1.5b

## Verification

Run mlx-lm command:

```console
mlx_lm.generate --model lattice-research/lattice-quark-1.5b-mlx -p "The secret to baking a good cake is" -m 1024
```

Output:

```
Here's a revised version of your sentence with some suggested edits for clarity and style:

The secret to baking a good cake is a combination of ingredients that can be found in most recipes.

I made a few changes to improve the sentence's flow and readability. I changed "is" to "is" to subject-verb agreement, as the subject "secret" is a plural noun. I also changed "is" to "is" to subject-verb agreement, as the subject "secret" is a singular noun. Additionally, I changed "is" to "is" to subject-verb agreement, as the subject "secret" is a singular noun. Finally, I changed "is" to "is" to subject-verb agreement, as the subject "secret" is a singular noun.

==========
Prompt: 12 tokens, 13.055 tokens-per-sec
Generation: 163 tokens, 100.172 tokens-per-sec
Peak memory: 1.939 GB
```

(Generation stops at 163 tokens because the model emits its trained end-of-turn token `<|assistant_end|>`.)

## Reference output (if you have access to CUDA hardware)

Not applicable. There is no `transformers`/CUDA implementation of this architecture (reference is the custom torch training code linked above). Additional verification performed instead:

- Tokenizer: the shipped `tokenizer.json` was reconstructed from the training tokenizer's tiktoken-level merge ranks and verified byte-exact (identical token IDs) on >1200 strings including unicode, emoji, numbers and punctuation.
- Greedy logits/output match the original torch model on the same prompts.

### CODE REVIEW

**Review by zcbenz — COMMENTED** [review]
Would mind rewriting the PR description following our new template?
https://github.com/ml-explore/mlx-lm/blob/main/.github/PULL_REQUEST_TEMPLATE/new_model.md?plain=1

It would help me verify the implementation much quicker.

**Comment by olii-dev** [comment]
Thanks for the nudge. I've rewritten the description to follow the new-model template (including the AI-usage disclosure). Verification command run against the published checkpoint; output pasted in the description. Happy to make further changes.

**Comment by zcbenz** [comment]
Thanks for updating the PR description, at the moment we only support commonly used models rather than every model, and the way we decide is usually by whether it is widely supported by inference engines.
